File: packages/modules/lodash/rust/src/object.rs
use std::collections::HashMap;

/// Copies own enumerable string keyed properties of source objects to the destination object.
pub fn assign(mut object: HashMap<String, String>, sources: &[HashMap<String, String>]) -> HashMap<String, String> {
    for source in sources {
        for (k, v) in source {
            object.insert(k.clone(), v.clone());
        }
    }
    object
}

/// Like `assign` but iterates over own and inherited source properties.
/// Implementation is identical to `assign` for `HashMap<String, String>`.
pub fn assign_in(mut object: HashMap<String, String>, sources: &[HashMap<String, String>]) -> HashMap<String, String> {
    for source in sources {
        for (k, v) in source {
            object.insert(k.clone(), v.clone());
        }
    }
    object
}

/// Like `assign_in` but accepts a customizer that determines the assigned value.
/// The customizer receives the existing value and the source value.
pub fn assign_in_with<F>(mut object: HashMap<String, String>, sources: &[HashMap<String, String>], customizer: F) -> HashMap<String, String>
where F: Fn(&str, &str) -> String
{
    for source in sources {
        for (k, v) in source {
            let existing = object.get(k).map(|s| s.as_str()).unwrap_or("");
            let new_value = customizer(existing, v);
            object.insert(k.clone(), new_value);
        }
    }
    object
}

/// Like `assign` but accepts a customizer that determines the assigned value.
/// The customizer receives the existing value and the source value.
pub fn assign_with<F>(mut object: HashMap<String, String>, sources: &[HashMap<String, String>], customizer: F) -> HashMap<String, String>
where F: Fn(&str, &str) -> String
{
    for source in sources {
        for (k, v) in source {
            let existing = object.get(k).map(|s| s.as_str()).unwrap_or("");
            let new_value = customizer(existing, v);
            object.insert(k.clone(), new_value);
        }
    }
    object
}

/// Creates a slice of values corresponding to `paths` of the object.
pub fn at<'a>(object: &'a HashMap<String, String>, paths: &[String]) -> Vec<Option<&'a String>> {
    paths.iter().map(|p| object.get(p)).collect()
}

/// Creates an object that inherits from the given prototype.
/// Returns the prototype directly since `HashMap` doesn't support prototypal inheritance.
pub fn create(prototype: HashMap<String, String>) -> HashMap<String, String> {
    prototype
}

/// Assigns own enumerable string keyed properties of source objects to the destination object
/// for all destination properties that resolve to undefined.
pub fn defaults(mut object: HashMap<String, String>, sources: &[HashMap<String, String>]) -> HashMap<String, String> {
    for source in sources {
        for (k, v) in source {
            if !object.contains_key(k) {
                object.insert(k.clone(), v.clone());
            }
        }
    }
    object
}

/// Like `defaults` but recursively assigns. Simplified to match `defaults` behavior.
pub fn defaults_deep(mut object: HashMap<String, String>, sources: &[HashMap<String, String>]) -> HashMap<String, String> {
    for source in sources {
        for (k, v) in source {
            if !object.contains_key(k) {
                object.insert(k.clone(), v.clone());
            }
        }
    }
    object
}

/// Finds the first key for which the predicate returns true.
pub fn find_key<F>(object: &HashMap<String, String>, predicate: F) -> Option<String>
where F: Fn(&str) -> bool
{
    for (k, v) in object {
        if predicate(v) {
            return Some(k.clone());
        }
    }
    None
}

/// Like `find_key` but iterates from right to left.
/// Identical to `find_key` since `HashMap` is unordered.
pub fn find_last_key<F>(object: &HashMap<String, String>, predicate: F) -> Option<String>
where F: Fn(&str) -> bool
{
    find_key(object, predicate)
}

/// Iterates over own and inherited enumerable string keyed properties of an object.
pub fn for_in<F>(object: &HashMap<String, String>, iteratee: F)
where F: Fn(&str, &str)
{
    for (k, v) in object {
        iteratee(k, v);
    }
}

/// Like `for_in` but iterates from right to left.
/// Identical to `for_in` since `HashMap` is unordered.
pub fn for_in_right<F>(object: &HashMap<String, String>, iteratee: F)
where F: Fn(&str, &str)
{
    for_in(object, iteratee);
}

/// Iterates over own enumerable string keyed properties of an object.
pub fn for_own<F>(object: &HashMap<String, String>, iteratee: F)
where F: Fn(&str, &str)
{
    for (k, v) in object {
        iteratee(k, v);
    }
}

/// Like `for_own` but iterates from right to left.
/// Identical to `for_own` since `HashMap` is unordered.
pub fn for_own_right<F>(object: &HashMap<String, String>, iteratee: F)
where F: Fn(&str, &str)
{
    for_own(object, iteratee);
}

/// Gets the function names of an object. Always returns an empty vec for `HashMap`.
pub fn functions(_object: &HashMap<String, String>) -> Vec<String> {
    vec![]
}

/// Gets the function names of an object, including inherited functions.
/// Always returns an empty vec for `HashMap`.
pub fn functions_in(_object: &HashMap<String, String>) -> Vec<String> {
    vec![]
}

/// Gets the value at `path` of `object`. If the value is not found, returns `None`.
/// The `default_value` parameter is accepted for API compatibility but cannot be
/// returned due to type constraints (`&str` vs `&String`).
pub fn get<'a>(object: &'a HashMap<String, String>, path: &str, _default_value: Option<&'a str>) -> Option<&'a String> {
    object.get(path)
}

/// Checks if `path` is a direct property of `object`.
pub fn has(object: &HashMap<String, String>, path: &str) -> bool {
    object.contains_key(path)
}

/// Checks if `path` is a direct or inherited property of `object`.
pub fn has_in(object: &HashMap<String, String>, path: &str) -> bool {
    object.contains_key(path)
}

/// Creates an object composed of the inverted keys and values of `object`.
/// If duplicate values are encountered, later keys overwrite earlier ones.
pub fn invert(object: &HashMap<String, String>) -> HashMap<String, String> {
    let mut result = HashMap::new();
    for (k, v) in object {
        result.insert(v.clone(), k.clone());
    }
    result
}

/// Like `invert` but the inverted keys are generated by running each value through `iteratee`.
pub fn invert_by<F>(object: &HashMap<String, String>, iteratee: F) -> HashMap<String, String>
where F: Fn(&str) -> String
{
    let mut result = HashMap::new();
    for (k, v) in object {
        let new_key = iteratee(v);
        result.insert(new_key, k.clone());
    }
    result
}

/// Invokes the method at `path` of `object`.
/// Simplified to just get the value at path.
pub fn invoke<'a>(object: &'a HashMap<String, String>, path: &str) -> Option<&'a String> {
    object.get(path)
}

/// Gets the own enumerable property names of `object`.
pub fn keys(object: &HashMap<String, String>) -> Vec<String> {
    object.keys().cloned().collect()
}

/// Gets the own and inherited enumerable property names of `object`.
pub fn keys_in(object: &HashMap<String, String>) -> Vec<String> {
    object.keys().cloned().collect()
}

/// Creates an object with the same values as `object` and keys generated by running
/// each own enumerable string keyed property through `iteratee`.
pub fn map_keys<F>(object: &HashMap<String, String>, iteratee: F) -> HashMap<String, String>
where F: Fn(&str) -> String
{
    let mut result = HashMap::new();
    for (k, v) in object {
        result.insert(iteratee(k), v.clone());
    }
    result
}

/// Creates an object with the same keys as `object` and values generated by running
/// each own enumerable string keyed property through `iteratee`.
pub fn map_values<F>(object: &HashMap<String, String>, iteratee: F) -> HashMap<String, String>
where F: Fn(&str) -> String
{
    let mut result = HashMap::new();
    for (k, v) in object {
        result.insert(k.clone(), iteratee(v));
    }
    result
}

/// Recursively merges own and inherited enumerable string keyed properties of source
/// objects into the destination object. Simplified to match `assign` behavior.
pub fn merge(mut object: HashMap<String, String>, sources: &[HashMap<String, String>]) -> HashMap<String, String> {
    for source in sources {
        for (k, v) in source {
            object.insert(k.clone(), v.clone());
        }
    }
    object
}

/// Like `merge` but accepts a customizer that determines the merged value.
/// The customizer receives the existing value and the source value.
pub fn merge_with<F>(mut object: HashMap<String, String>, sources: &[HashMap<String, String>], customizer: F) -> HashMap<String, String>
where F: Fn(&str, &str) -> String
{
    for source in sources {
        for (k, v) in source {
            let existing = object.get(k).map(|s| s.as_str()).unwrap_or("");
            let new_value = customizer(existing, v);
            object.insert(k.clone(), new_value);
        }
    }
    object
}

/// Creates an object composed of the own enumerable property paths of `object`
/// that are not omitted.
pub fn omit(object: &HashMap<String, String>, paths: &[String]) -> HashMap<String, String> {
    let mut result = HashMap::new();
    for (k, v) in object {
        if !paths.contains(k) {
            result.insert(k.clone(), v.clone());
        }
    }
    result
}

/// Creates an object composed of the own enumerable property paths of `object`
/// that are not omitted, excluding properties for which the predicate returns truthy.
pub fn omit_by<F>(object: &HashMap<String, String>, predicate: F) -> HashMap<String, String>
where F: Fn(&str) -> bool
{
    let mut result = HashMap::new();
    for (k, v) in object {
        if !predicate(k) {
            result.insert(k.clone(), v.clone());
        }
    }
    result
}

/// Creates an object composed of the picked `paths` of `object`.
pub fn pick(object: &HashMap<String, String>, paths: &[String]) -> HashMap<String, String> {
    let mut result = HashMap::new();
    for path in paths {
        if let Some(v) = object.get(path) {
            result.insert(path.clone(), v.clone());
        }
    }
    result
}

/// Creates an object composed of the own enumerable property paths of `object`
/// for which the predicate returns truthy.
pub fn pick_by<F>(object: &HashMap<String, String>, predicate: F) -> HashMap<String, String>
where F: Fn(&str) -> bool
{
    let mut result = HashMap::new();
    for (k, v) in object {
        if predicate(k) {
            result.insert(k.clone(), v.clone());
        }
    }
    result
}

/// Gets the value at `path` of `object`. Simplified: same as `get`.
pub fn result<'a>(object: &'a HashMap<String, String>, path: &str) -> Option<&'a String> {
    object.get(path)
}

/// Sets the value at `path` of `object`.
pub fn set(object: &mut HashMap<String, String>, path: &str, value: String) {
    object.insert(path.to_string(), value);
}

/// Like `set` but accepts a customizer that determines the value to set.
/// The customizer receives the existing value (or "") and the new value.
pub fn set_with<F>(object: &mut HashMap<String, String>, path: &str, value: String, customizer: F)
where F: Fn(&str, &str) -> String
{
    let existing = object.get(path).map(|s| s.as_str()).unwrap_or("");
    let new_value = customizer(existing, &value);
    object.insert(path.to_string(), new_value);
}

/// Creates a vector of key-value pairs for `object`.
pub fn to_pairs(object: &HashMap<String, String>) -> Vec<(String, String)> {
    object.iter().map(|(k, v)| (k.clone(), v.clone())).collect()
}

/// Creates a vector of key-value pairs for `object`, including inherited properties.
pub fn to_pairs_in(object: &HashMap<String, String>) -> Vec<(String, String)> {
    object.iter().map(|(k, v)| (k.clone(), v.clone())).collect()
}

/// An alternative to a reducer for objects, transforming an object into a new accumulator.
pub fn transform<F>(object: &HashMap<String, String>, iteratee: F, mut accumulator: HashMap<String, String>) -> HashMap<String, String>
where F: Fn(&str, &str, &mut HashMap<String, String>)
{
    for (k, v) in object {
        iteratee(k, v, &mut accumulator);
    }
    accumulator
}

/// Removes the property at `path` of `object`. Returns true if the property was deleted.
pub fn unset(object: &mut HashMap<String, String>, path: &str) -> bool {
    object.remove(path).is_some()
}

/// Updates the value at `path` of `object` using the `updater` function.
/// The updater receives the current value (or `None` if not present) and returns the new value.
pub fn update<F>(object: &mut HashMap<String, String>, path: &str, updater: F)
where F: Fn(Option<&str>) -> String
{
    let current = object.get(path).map(|s| s.as_str());
    let new_value = updater(current);
    object.insert(path.to_string(), new_value);
}

/// Like `update` but accepts a customizer that transforms the result of the updater
/// before setting it. The customizer receives the existing value and the updater result.
pub fn update_with<F, G>(object: &mut HashMap<String, String>, path: &str, updater: F, customizer: G)
where F: Fn(Option<&str>) -> String, G: Fn(&str, &str) -> String
{
    let current = object.get(path).map(|s| s.as_str());
    let updated = updater(current);
    let existing = object.get(path).map(|s| s.as_str()).unwrap_or("");
    let final_value = customizer(existing, &updated);
    object.insert(path.to_string(), final_value);
}

/// Gets the own enumerable string keyed property values of `object`.
pub fn values(object: &HashMap<String, String>) -> Vec<String> {
    object.values().cloned().collect()
}

/// Gets the own and inherited enumerable string keyed property values of `object`.
pub fn values_in(object: &HashMap<String, String>) -> Vec<String> {
    object.values().cloned().collect()
}

File: packages/modules/lodash/core/python/libs/object.py
"""
object
"""

import copy

from .lang import is_function


def assign(obj, *args):
    """
    assign
    """
    for arg in args:
        for key in arg:
            value = arg[key]
            obj[key] = value
    return obj


def at(object_var, paths):
    """
    at
    """
    array = []
    for path in paths:
        value = get(object_var, path)
        array.append(value)
    return array


def defaults(*args):
    """
    defaults
    """
    objects = list(args)
    objects.reverse()
    object_var = {}
    for item in objects:
        for key in item.keys():
            value = item[key]
            object_var[key] = value
    return object_var


def find_key(object_var, callback):
    """
    find_key
    """
    _keys = keys(object_var)
    new_keys = []
    for key in _keys:
        value = object_var[key]
        if callback(value):
            new_keys.append(key)
    if len(new_keys) > 0:
        return new_keys[0]
    return None


def find_last_key(object_var, callback):
    """
    find_last_key
    """
    _keys = keys(object_var)
    _keys.reverse()
    new_keys = []
    for key in _keys:
        value = object_var[key]
        if callback(value):
            new_keys.append(key)
    if len(new_keys) > 0:
        return new_keys[0]
    return None


def for_in(object_var, callback):
    """
    for_in
    """
    _keys = keys(object_var)
    for key in _keys:
        value = object_var[key]
        callback(value, key)


def for_in_right(object_var, callback):
    """
    for_in_right
    """
    _keys = keys(object_var)
    _keys.reverse()
    for key in _keys:
        value = object_var[key]
        callback(value, key)


def for_own(object_var, callback):
    """
    for_own
    """
    for_in(object_var, callback)


def for_own_right(object_var, callback):
    """
    for_own_right
    """
    for_in_right(object_var, callback)


def functions(object_var):
    """
    functions
    """
    _keys = keys(object_var)
    new_array = []
    for key in _keys:
        value = object_var[key]
        new_array.append(value)
    return new_array


def functions_in(object_var):
    """
    functions_in
    """
    return functions(object_var)


def get(obj, path, default_value=None):
    """
    get
    """
    has_path = has(obj, path)
    if not has_path:
        return default_value
    _keys = path.split(".")
    value = obj
    for key in _keys:
        value = value.get(key)
    return value


def has(obj, path):
    """
    has
    """
    _keys = path.split(".")
    flag = True
    nested = obj
    for key in _keys:
        if not isinstance(nested, dict):
            flag = False
            break
        if key in nested:
            nested = nested[key]
        else:
            flag = False
            break
    return flag


def invert(obj):
    """
    invert
    """
    new_obj = {}
    _keys = keys(obj)
    for key in _keys:
        value = obj[key]
        new_key = str(value)
        new_obj[new_key] = key
    return new_obj


def invert_by(obj, callback=None):
    """
    invert_by
    """
    new_obj = {}
    _keys = keys(obj)
    for key in _keys:
        value = obj[key]
        new_key = str(value)
        if is_function(callback):
            new_key = callback(value)
        if new_key in new_obj:
            new_obj[new_key].append(key)
        else:
            new_obj[new_key] = [key]
    return new_obj


def keys(obj):
    """
    keys
    """
    return list(obj.keys())


def keys_in(obj):
    """
    keys_in
    """
    return list(obj.keys())


def map_keys(obj, callback):
    """
    map_keys
    """
    old_keys = keys(obj)
    new_obj = {}
    for key in old_keys:
        value = obj[key]
        new_key = callback(value, key)
        new_obj[new_key] = value
    return new_obj


def map_values(object_var, callback):
    """
    map_values
    """
    _keys = keys(object_var)
    for key in _keys:
        value = object_var[key]
        new_value = callback(value)
        object_var[key] = new_value
    return object_var


def omit(object_var, paths):
    """
    omit
    """
    new_object = {}
    _keys = keys(object_var)
    for key in _keys:
        if key not in paths:
            value = object_var[key]
            new_object[key] = value
    return new_object


def omit_by(obj, callback):
    """
    omit_by
    """
    new_object = {}
    _keys = keys(obj)
    for key in _keys:
        value = obj[key]
        flag = callback(value)
        if not flag:
            new_object[key] = value
    return new_object


def pick(obj, paths):
    """
    pick
    """
    new_obj = {}
    for path in paths:
        value = get(obj, path)
        set_object(new_obj, path, value)
    return new_obj


def pick_by(obj, callback):
    """
    pick_by
    """
    new_obj = {}
    _keys = keys(obj)
    for key in _keys:
        value = get(obj, key)
        flag = callback(value)
        if flag:
            set_object(new_obj, key, value)
    return new_obj


def result(obj, path, default_value=None):
    """
    result
    """
    callback = get(obj, path, default_value)
    if not callable(callback):
        return callback
    value = callback()
    if value is None:
        return default_value
    return value


def set_object(obj, path, value):
    """
    set_object
    """
    _keys = path.split(".")
    if len(_keys) == 1:
        obj[_keys[0]] = value
    else:
        key = _keys[0]
        object_value = set_object({}, ".".join(_keys[1 : len(_keys)]), value)
        obj[key] = object_value
    return obj


def to_pairs(obj):
    """
    to_pairs
    """
    array = []
    for key in obj:
        value = obj[key]
        array.append([key, value])
    return array


def to_pairs_in(obj):
    """
    to_pairs_in
    """
    return to_pairs(obj)


def update(obj, path, updater):
    """
    update
    """
    old_value = get(obj, path)
    if old_value is None:
        return obj
    new_value = updater(old_value)
    new_obj = set_object(obj, path, new_value)
    return new_obj


def values(obj):
    """
    values
    """
    return list(obj.values())


def values_in(obj):
    """
    values_in
    """
    return values(obj)


def assign_in(obj, *args):
    """
    assign_in
    """
    return assign(obj, *args)


def assign_with(obj, *args, customizer=None):
    """
    assign_with
    """
    if customizer is None:
        return assign(obj, *args)
    for arg in args:
        for key in arg:
            value = arg[key]
            new_value = customizer(obj.get(key), value, key, obj, arg)
            if new_value is not None:
                obj[key] = new_value
            else:
                obj[key] = value
    return obj


def assign_in_with(obj, *args, customizer=None):
    """
    assign_in_with
    """
    return assign_with(obj, *args, customizer=customizer)


def create(prototype, props=None):
    """
    create
    """
    obj = {}
    if isinstance(prototype, dict):
        for key in prototype:
            obj[key] = prototype[key]
    if props is not None:
        for key in props:
            obj[key] = props[key]
    return obj


def defaults_deep(*args):
    """
    defaults_deep
    """
    objects = list(args)
    objects.reverse()
    result = {}
    for item in objects:
        for key in item.keys():
            if key not in result:
                result[key] = copy.deepcopy(item[key])
            else:
                existing = result[key]
                incoming = item[key]
                if isinstance(existing, dict) and isinstance(incoming, dict):
                    result[key] = defaults_deep(existing, incoming)
    return result


def has_in(obj, path):
    """
    has_in
    """
    return has(obj, path)


def get_value(obj, path):
    """
    get_value
    """
    _keys = path.split(".")
    value = obj
    for key in _keys:
        if isinstance(value, dict) and key in value:
            value = value[key]
        else:
            return None
    return value


def invoke(obj, path, *args):
    """
    invoke
    """
    _keys = path.split(".")
    method_name = _keys[-1]
    target = obj
    for key in _keys[:-1]:
        if isinstance(target, dict) and key in target:
            target = target[key]
        else:
            return None
    method = getattr(target, method_name, None)
    if not callable(method):
        return None
    return method(*args)


def merge(obj, *sources):
    """
    merge
    """
    for source in sources:
        for key in source:
            if key in obj and isinstance(obj[key], dict) and isinstance(source[key], dict):
                obj[key] = merge(obj[key], source[key])
            else:
                obj[key] = copy.deepcopy(source[key])
    return obj


def merge_with(obj, *sources, customizer=None):
    """
    merge_with
    """
    if customizer is None:
        return merge(obj, *sources)
    for source in sources:
        for key in source:
            result_value = customizer(obj.get(key), source[key], key, obj, source)
            if result_value is not None:
                obj[key] = result_value
            elif key in obj and isinstance(obj[key], dict) and isinstance(source[key], dict):
                obj[key] = merge_with(obj[key], source[key], customizer=customizer)
            else:
                obj[key] = copy.deepcopy(source[key])
    return obj


def set_with(obj, path, value, customizer=None):
    """
    set_with
    """
    _keys = path.split(".")
    if customizer is not None:
        nested = obj
        for i, key in enumerate(_keys[:-1]):
            if key not in nested:
                nested[key] = {}
            nested = nested[key]
        last_key = _keys[-1]
        new_value = customizer(nested.get(last_key), value, last_key, obj, path)
        nested[last_key] = new_value if new_value is not None else value
        return obj
    return set_object(obj, path, value)


def transform(obj, callback, accumulator=None):
    """
    transform
    """
    if accumulator is None:
        accumulator = {}
    _keys = keys(obj)
    for key in _keys:
        value = obj[key]
        callback(accumulator, value, key)
    return accumulator


def unset(obj, path):
    """
    unset
    """
    _keys = path.split(".")
    nested = obj
    for key in _keys[:-1]:
        if not isinstance(nested, dict) or key not in nested:
            return False
        nested = nested[key]
    last_key = _keys[-1]
    if last_key in nested:
        del nested[last_key]
        return True
    return False


def update_with(obj, path, updater, customizer=None):
    """
    update_with
    """
    _keys = path.split(".")
    value = get_value(obj, path)
    if value is None:
        return obj
    new_value = updater(value)
    if customizer is not None:
        return set_with(obj, path, new_value, customizer=customizer)
    return set_object(obj, path, new_value)
